utils/pubsub.go

```diff
@@ -14,56 +14,108 @@ type DataEvent struct {
 // DataChannel is a channel which can accept an DataEvent
 type DataChannel chan DataEvent
 
-// DataChannelSlice is a slice of DataChannels
-type DataChannelSlice []DataChannel
 type PublishSubscriber interface {
 	Publish(topic string, data interface{})
-	PublishToChannel(channel DataChannel, topic string, data interface{})
 	Subscribe(topic string, ch DataChannel)
 }
 
 // EventBus stores the information about subscribers interested for a particular topic
 type EventBus struct {
-	subscribers map[string]DataChannelSlice
-	rm          sync.RWMutex
+	lastEventMutex sync.RWMutex
+	// lastEvent holds the last event for each topic so that we can send it to new subscribers
+	lastEvent map[string]*DataEvent
+
+	subscriptionsMutex sync.RWMutex
+	// subscriptions keep the list of subscription publishers per topic
+	subscriptions map[string]subPublishers
 }
 
 func (eb *EventBus) Publish(topic string, data interface{}) {
-	eb.rm.RLock()
-	if chans, found := eb.subscribers[topic]; found {
-		// this is done because the slices refer to same array even though they are passed by value
-		// thus we are creating a new slice with our elements thus preserve locking correctly.
-		// special thanks for /u/freesid who pointed it out
-		channels := append(DataChannelSlice{}, chans...)
-		go func(data DataEvent, dataChannelSlices DataChannelSlice) {
-			for _, ch := range dataChannelSlices {
-				//Publishing to channels in separate goroutines
-				go func(ch DataChannel, data DataEvent) {
-					ch <- data
-				}(ch, data)
-			}
-		}(DataEvent{Data: data, Topic: topic}, channels)
+	eb.subscriptionsMutex.RLock()
```

Comment:
Do you really need to start locking here given that you access `eb.subscriptions` only in line 45?

Reply:
I am trying to keep a consistent lock order between the two locks in order to avoid a deadlock. The same locks are acquired by `Subscribe` in line 52